### DISCUSSION TITLE
Sonoff POW R2 delayed power off via built-in button

### DISCUSSION THREAD

**Discussion by JMVS** [arendst/Tasmota#13926] [Q&A] [answered]
Hello.

Let me explain what I'm trying to achieve to see if it possible to do or provide alternatives.

Right now I have a **Sonoff POW R2** with Tasmota _(at the time of writing this using version 10.0.0)_ connected to a **3D Printer** _(with printer's switch always on)_ and a **RasperryPi** (running OctoPrint) connected to the printer's PSU _(meaning that if I power off the printer, it also powers off the RPi)_. As many will know, the RPi needs to be shut off cleanly before removing power from it. I also can't add a switch to the POW due to [hardware limitations](https://tasmota.github.io/docs/devices/Sonoff-Pow-R2/#special-attention).
As a sidenote, I also have Home Assistant running on a home server.
Right now I can shutdown the RPi cleanly from the LCD's printer and then power everything off via the POW R2 button.

What I would like to achieve is setting a way for **delaying the power off the POW R2 for around 30sec** after pressing the button _(or whatever other means like webinterface and/or Home Assistant)_. This gives me time to shutdown the RPi cleanly and all could be done via the printer's LCD menu _(which is outside the scope of this question)_. 

I've been trying to implement this with [these rules](https://tasmota.github.io/docs/Rules/#time-delay-after-switch-off) altering swtich for button and playing with the Switchmodes but no luck so far.

I'm positive that I can achieve this using Home Assistant but I'm trying to do this as standalone as possible.

Thanks in advance.

**Comment by sfromis** [comment]
In general, you can take over the button using [rules](https://tasmota.github.io/docs/Rules/) (and possibly SetOption73, not required for this), and do any action you see fit, like starting a timer to later switch power off. Using SwitchMode only works when the input is configured as a Switch, not as a Button. You could change the config, if that suits you. 

As an example, you can have rules like:
```
on Button1#state==1 do Power on endon
on Button1#state==0 do RuleTimer1 30 endon
on Rules#Timer==1 do Power off endon
```

**Comment by barbudor** [comment, marked as answer]
A MQTT POWER command, or pressing hthe Web button will always have immediate effect. You can't delay that.

What you can do is starting a RuleTimer on
- a Button press, but a button doesn't know when power is on or off. SO you will need some logic to have immediate on and delayed off
- an Event command (like instead of sending `cmnd/xxx/power off`, your external system would send `cmnd/xxx/event mypower=0`
- When power consumption drop below a certain threshold 

And you can combine those if you want

@sfromis, Button#state doesn't have 0 or 1 value. Only 2 (toggle), 4 (hold) or multipress

**Reply by JMVS** [reply]
Thank you for the suggestions

> A MQTT POWER command, or pressing hthe Web button will always have immediate effect. You can't delay that.
> 
> What you can do is starting a RuleTimer on
> 
>     * a Button press, but a button doesn't know when power is on or off. SO you will need some logic to have immediate on and delayed off
> 
>     * an Event command (like instead of sending `cmnd/xxx/power off`, your external system would send `cmnd/xxx/event mypower=0`
> 
>     * When power consumption drop below a certain threshold
> 
> 
> And you can combine those if you want
> .
> @sfromis, Button#state doesn't have 0 or 1 value. Only 2 (toggle), 4 (hold) or multipress

Your third suggestion seems the closest to what I want to achieve. Do you have any documentation of example I can read and adapt?

**Reply by barbudor** [reply]
This is what I'm using on plugs connected to PCs so power is cut a few seconds after they shutdown
```
Rule1 on Energy#Power<%Mem1% do ruletimer1 %Mem3% endon  on Energy#Power>%Mem2% do ruletimer1 0 endon
Rule1 1
Rule1 5
Rule2 on rules#timer=1 do power off endon
Rule2 1
```
With:
- `mem1` is power level to start the timer when power drop below this value
- `mem2` is power level to clear the timer if power ever goes over this value. Must be greater than mem1. generally I set it 1 or 2 W above mem1
- `mem3` is the timer duration
Note that Rule1 must be set in "once" mode with `Rule1 5`

As I said, pressing the web button or sending `POWER off` over MQTT will always have an immediate effect. Physical button on the POW R2 can be disabled wiht a rule if needed.

**Reply by JMVS** [reply]
Thank you @barbudor! This works even better than I initially wanted to do. I've tested it and it's working great so far.
